File: module_base/src/main/java/com/app/base/data/NetProviderImpl.java
package com.app.base.data;

import com.android.base.utils.android.ResourceUtils;
import com.android.sdk.net.core.ExceptionFactory;
import com.android.sdk.net.exception.ApiErrorException;
import com.android.sdk.net.provider.ApiHandler;
import com.android.sdk.net.provider.ErrorDataAdapter;
import com.android.sdk.net.provider.ErrorMessage;
import com.android.sdk.net.provider.HttpConfig;
import com.android.sdk.net.provider.PostTransformer;
import com.app.base.BuildConfig;
import com.app.base.R;
import com.app.base.data.api.ApiHelper;
import com.app.base.debug.DebugTools;

import org.reactivestreams.Publisher;

import java.lang.annotation.Annotation;
import java.lang.reflect.Type;
import java.util.concurrent.TimeUnit;

import io.reactivex.Flowable;
import io.reactivex.Observable;
import io.reactivex.ObservableSource;
import io.reactivex.Single;
import io.reactivex.SingleSource;
import okhttp3.OkHttpClient;
import okhttp3.ResponseBody;
import okhttp3.logging.HttpLoggingInterceptor;
import retrofit2.Retrofit;
import timber.log.Timber;

/**
 * @author Ztiany
 * Email: ztiany3@gmail.com
 * Date : 2018-11-08 16:16
 */
final class NetProviderImpl {

    ExceptionFactory mExceptionFactory = result -> null;

    ApiHandler mApiHandler = result -> {
        int code = result.getCode();
        //登录状态已过期，请重新登录、账号在其他设备登陆
        if (ApiHelper.isLoginExpired(code)) {
            DataContext.getInstance().publishLoginExpired(code);
        }
    };

    HttpConfig mHttpConfig = new HttpConfig() {

        private static final int CONNECTION_TIME_OUT = 10;
        private static final int IO_TIME_OUT = 20;

        @Override
        public void configHttp(OkHttpClient.Builder builder) {
            //常规配置
            builder.connectTimeout(CONNECTION_TIME_OUT, TimeUnit.SECONDS)
                    .readTimeout(IO_TIME_OUT, TimeUnit.SECONDS)
                    .writeTimeout(IO_TIME_OUT, TimeUnit.SECONDS)
                    .hostnameVerifier((hostname, session) -> {
                        Timber.d("hostnameVerifier called with: hostname 、session = [" + hostname + "、" + session.getProtocol() + "]");
                        return true;
                    });

            //调试配置
            configDebug(builder);
        }

        void configDebug(OkHttpClient.Builder builder) {
            if (BuildConfig.openDebug) {
                HttpLoggingInterceptor httpLoggingInterceptor = new HttpLoggingInterceptor(
                        message -> Timber.tag("===OkHttp===").i(message)
                );
                httpLoggingInterceptor.setLevel(HttpLoggingInterceptor.Level.BODY);
                builder.addInterceptor(httpLoggingInterceptor);
                DebugTools.installStethoHttp(builder);
            }
        }

        @Override
        public boolean configRetrofit(Retrofit.Builder builder) {
            return false;
        }

        @Override
        public String baseUrl() {
            return DataContext.baseUrl();
        }
    };

    ErrorMessage mErrorMessage = new ErrorMessage() {
        @Override
        public CharSequence netErrorMessage(Throwable exception) {
            return ResourceUtils.getString(R.string.error_net_error);
        }

        @Override
        public CharSequence serverDataErrorMessage(Throwable exception) {
            return ResourceUtils.getString(R.string.error_service_data_error);
        }

        @Override
        public CharSequence serverErrorMessage(Throwable exception) {
            return ResourceUtils.getString(R.string.error_service_error);
        }

        @Override
        public CharSequence clientRequestErrorMessage(Throwable exception) {
            return ResourceUtils.getString(R.string.error_request_error);
        }

        @Override
        public CharSequence apiErrorMessage(ApiErrorException exception) {
            return ResourceUtils.getString(R.string.error_api_code_mask_tips, exception.getCode());
        }

        @Override
        public CharSequence unknowErrorMessage(Throwable exception) {
            return ResourceUtils.getString(R.string.error_unknow);
        }

    };

    ErrorDataAdapter mErrorDataAdapter = new ErrorDataAdapter() {
        @Override
        public Object createErrorDataStub(Type type, Annotation[] annotations, Retrofit retrofit, ResponseBody value) {
            return ApiHelper.newErrorDataStub();
        }

        @Override
        public boolean isErrorDataStub(Object object) {
            return ApiHelper.isDataError(object);
        }
    };

    PostTransformer mPostTransformer = new PostTransformer<Object>() {
        @Override
        public SingleSource<Object> apply(Single<Object> upstream) {
            return upstream;
        }

        @Override
        public Publisher<Object> apply(Flowable<Object> upstream) {
            return upstream;
        }

        @Override
        public ObservableSource<Object> apply(Observable<Object> upstream) {
            return upstream;
        }
    };

}

File: module_base/src/main/java/com/app/base/data/NetProviderImpl.kt
package com.app.base.data

import com.android.base.utils.android.views.getString
import com.android.sdk.net.core.exception.ApiErrorException
import com.android.sdk.net.core.provider.*
import com.android.sdk.net.rxjava.RxResultPostTransformer
import com.app.base.BuildConfig
import com.app.base.R
import com.app.base.data.api.ApiHelper
import com.app.base.debug.DebugTools
import io.reactivex.*
import okhttp3.OkHttpClient
import okhttp3.ResponseBody
import okhttp3.logging.HttpLoggingInterceptor
import org.reactivestreams.Publisher
import retrofit2.Retrofit
import timber.log.Timber
import java.lang.reflect.Type
import java.util.concurrent.TimeUnit
import javax.net.ssl.HostnameVerifier
import javax.net.ssl.SSLSession


internal fun newHttpConfig(): HttpConfig {

    return object : HttpConfig {

        private val CONNECTION_TIME_OUT = 10
        private val IO_TIME_OUT = 20

        override fun baseUrl() = DataContext.baseAPIUrl()
        override fun configRetrofit(okHttpClient: OkHttpClient, builder: Retrofit.Builder) = false

        override fun configHttp(builder: OkHttpClient.Builder) {
            //常规配置
            builder.connectTimeout(CONNECTION_TIME_OUT.toLong(), TimeUnit.SECONDS)
                    .readTimeout(IO_TIME_OUT.toLong(), TimeUnit.SECONDS)
                    .writeTimeout(IO_TIME_OUT.toLong(), TimeUnit.SECONDS)
                    .hostnameVerifier(HostnameVerifier { hostname, session ->
                        Timber.d("hostnameVerifier called with: hostname ${hostname}、session $session.protocol ")
                        true
                    })

            //调试配置
            if (BuildConfig.openDebug) {
                val httpLoggingInterceptor = HttpLoggingInterceptor(object : HttpLoggingInterceptor.Logger {
                    override fun log(message: String) {
                        Timber.tag("===OkHttp===").i(message)
                    }
                })
                httpLoggingInterceptor.level = HttpLoggingInterceptor.Level.BODY
                builder.addInterceptor(httpLoggingInterceptor)
                DebugTools.installStethoHttp(builder)
            }
        }
    }

}

internal fun newErrorMessage(): ErrorMessage {
    return object : ErrorMessage {
        override fun netErrorMessage(exception: Throwable): CharSequence {
            return getString(R.string.error_net_error)
        }

        override fun serverDataErrorMessage(exception: Throwable): CharSequence {
            return getString(R.string.error_service_data_error)
        }

        override fun serverErrorMessage(exception: Throwable): CharSequence {
            return getString(R.string.error_service_error)
        }

        override fun clientRequestErrorMessage(exception: Throwable): CharSequence {
            return getString(R.string.error_request_error)
        }

        override fun apiErrorMessage(exception: ApiErrorException): CharSequence {
            return getString(R.string.error_api_code_mask_tips, exception.code)
        }

        override fun unknowErrorMessage(exception: Throwable): CharSequence {
            return getString(R.string.error_unknow)
        }
    }
}

internal fun newErrorDataAdapter(): ErrorDataAdapter = object : ErrorDataAdapter {
    override fun createErrorDataStub(type: Type, annotations: Array<Annotation>, retrofit: Retrofit, value: ResponseBody): Any {
        return ApiHelper.newErrorDataStub()
    }

    override fun isErrorDataStub(`object`: Any): Boolean {
        return ApiHelper.isDataError(`object`)
    }
}

internal fun newPostTransformer(): RxResultPostTransformer<*> = object : RxResultPostTransformer<Any> {
    override fun apply(upstream: Single<Any>): SingleSource<Any> {
        return upstream
    }

    override fun apply(upstream: Flowable<Any>): Publisher<Any> {
        return upstream
    }

    override fun apply(upstream: Observable<Any>): ObservableSource<Any> {
        return upstream
    }
}

internal fun newApiHandler(): ApiHandler = ApiHandler { result ->
    //登录状态已过期，请重新登录、账号在其他设备登陆
    if (ApiHelper.isLoginExpired(result.code)) {
        DataContext.getInstance().publishLoginExpired(ApiErrorException(result.code, "登录过期"))
    }
}